sdk/storage/azblob/internal/exported/blob_batch.go

```diff
@@ -0,0 +1,269 @@
+
+	// Close the multipart writer
+	err = writer.Close()
+	if err != nil {
+		return nil, "", err
+	}
+
+	return reqBody.Bytes(), batchID, nil
+}
+
+// UpdateSubRequestHeaders updates the sub-request headers.
+// Removes x-ms-version header.
+func UpdateSubRequestHeaders(req *policy.Request) {
+	// remove x-ms-version header from the request header
+	for k := range req.Raw().Header {
+		if strings.EqualFold(k, shared.HeaderXmsVersion) {
+			delete(req.Raw().Header, k)
+		}
+	}
+}
+
+// BlobBatchSubResponse contains the response for the individual sub-requests.
+type BlobBatchSubResponse struct {
+	ContentID     *int
+	ContainerName *string
+	BlobName      *string
+	RequestID     *string
+	Version       *string
+	Error         error // nil error indicates that the batch sub-request operation is successful
+}
+
+func getResponseBoundary(contentType *string) (string, error) {
+	if contentType == nil {
+		return "", fmt.Errorf("Content-Type returned in SubmitBatch response is nil")
+	}
+	boundaryIdx := strings.Index(*contentType, "batchresponse_")
+	if boundaryIdx == -1 {
+		return "", fmt.Errorf("batch boundary not present in Content-Type header of the SubmitBatch response.\nContent-Type: %v\n", *contentType)
+	}
+	return (*contentType)[boundaryIdx:], nil
+}
+
+func getContentID(part *multipart.Part) (*int, error) {
+	contentID := getResponseHeader("Content-ID", part.Header)
+	if contentID == nil {
+		return nil, nil
+	}
+
+	val, err := strconv.Atoi(strings.TrimSpace(*contentID))
+	if err != nil {
+		return nil, err
+	}
+	return &val, nil
+}
+
+func getResponseHeader(key string, headers map[string][]string) *string {
+	for k, v := range headers {
+		if strings.EqualFold(k, key) {
+			return to.Ptr(v[0])
+		}
+	}
+	return nil
+}
+
+func ParseBlobBatchResponse(respBody io.ReadCloser, contentType *string, subRequests []*policy.Request) ([]*BlobBatchSubResponse, error) {
+	boundary, err := getResponseBoundary(contentType)
+	if err != nil {
+		return nil, err
+	}
+
+	respReader := multipart.NewReader(respBody, boundary)
+	var responses []*BlobBatchSubResponse
+
+	for {
+		part, err := respReader.NextPart()
+		if err != nil {
+			if err == io.EOF {
+				break
+			} else {
+				return nil, err
+			}
+		}
+
+		batchSubResponse := &BlobBatchSubResponse{}
+		batchSubResponse.ContentID, err = getContentID(part)
+		if err != nil {
+			return nil, err
+		}
+
+		if batchSubResponse.ContentID != nil {
+			path := strings.Trim(subRequests[*batchSubResponse.ContentID].Raw().URL.Path, "/")
+			p := strings.Split(path, "/")
+			batchSubResponse.ContainerName = to.Ptr(p[0])
+			batchSubResponse.BlobName = to.Ptr(strings.Join(p[1:], "/"))
+		}
+
+		respBytes, err := ioutil.ReadAll(part)
+		if err != nil {
+			return nil, err
+		}
+		respBytes = append(respBytes, byte('\n'))
+		buf := bytes.NewBuffer(respBytes)
+		resp, err := http.ReadResponse(bufio.NewReader(buf), nil)
+		// sub-response parsing error
+		if err != nil {
+			return nil, err
+		}
+
+		if resp != nil {
```

Comment:
Would this ever be `nil` if `err` above is not `nil`?  If it was `nil`, then it would appear that there would be an empty `BlobBatchSubResponse` entry in the slice.

Reply:
Removed the `nil` check for `resp`